internal/pkg/agent/application/upgrade/marker_watcher.go

```diff
@@ -115,15 +115,18 @@ func (mfw *MarkerFileWatcher) processMarker(currentVersion string) {
 	}
 
 	// If the marker exists but the version of Agent we're running right
-	// now is the same as the prevVersion recorded in the marker, it means
-	// the upgrade was rolled back. Ideally, this UPG_ROLLBACK state would've
-	// been recorded in the marker's upgrade details field but, in case it
-	// isn't for some reason, we fallback to explicitly setting that state as
-	// part of the upgrade details in the marker.
+	// now is the same as the prevVersion recorded in the marker, it could mean
+	// that we are in one of two states:
+	// - the marker was just created and the upgrade is about to start
+	//   (marker.details.state should not be empty), or
+	// - the upgrade was rolled back (marker.details.state should be UPG_ROLLBACK)
```

Comment:
Is there a third option we haven't considered? The case where the agent is restarted in the middle of an upgrade, for example when it is in the UPGRADE_EXTRACTING state. The obvious example would be the host system powering off.

What happens in this case? What is reported in the upgrade details?

Reply:
If we just rolled back, is the previous version of the agent guaranteed to be next to the currently running agent in the data path? That is is the state of the filesystem always something like:

```
data/
  elastic-agent-current/
  elastic-agent-next/
```

If this is true, can we look to see if there's another agent in the data directory next to us to detect if there was a roll back?

Reply:
We will likely have both versions next to each other if the upgrade is interrupted immediately after the artifact is extracted, so that may not be 100% either.

One more thing, in the case that the host system powers off we should start calling https://pkg.go.dev/os#File.Sync on the marker file. It doesn't look like we do this today.

https://github.com/elastic/elastic-agent/blob/97e82175953af75fbe29fccc3cfcc5688a96dc8f/internal/pkg/agent/application/upgrade/marker_access_other.go#L26-L30

This might only be worth doing arounds critical transitions, like when the agent process re-execs or the watcher first starts up.

Reply:
I've implemented the fsync change in its own PR, since it's not strictly related to this PR here: https://github.com/elastic/elastic-agent/pull/3836

Reply:
> Is there a third option we haven't considered? The case where the agent is restarted in the middle of an upgrade, for example when it is in the UPGRADE_EXTRACTING state. The obvious example would be the host system powering off.

> What happens in this case? What is reported in the upgrade details?

It depends on whether the state transition happens before or after the upgrade marker file comes into existence.  

The following states occur before the upgrade marker file comes into existence and, as such, are never persisted in it: `UPG_REQUESTED`, `UPG_SCHEDULED`, `UPG_DOWNLOADING`, `UPG_EXTRACTING`.   Additionally, the `UPG_RESTARTING` state is also currently not being persisted to the upgrade marker file, mostly because of where this state transition happens in the code vs. where the upgrade marker file is being created. With some refactoring, we could start persisting this state to the upgrade marker file as well. So if the Agent were to restart during one of these states, the upgrade details that are stored in the Coordinator state (and from there sent to Fleet) would get reset to nothing and the upgrade state would be lost.

The following states occur right before or after the upgrade marker file comes into existence and, as such, do get persisted to it: `UPG_REPLACING`, `UPG_WATCHING`, `UPG_ROLLBACK`.  So if the Agent were to restart during one of these states, the upgrade details from the upgrade marker would be restored to the Coordinator state (and from there sent to Fleet).

We may want to consider either persisting upgrade details in their own file throughout the upgrade process OR creating the upgrade marker file at the start of the upgrade process instead of where it's being created now (right before the Upgrade Watcher is invoked from the old Agent).

Reply:
> If we just rolled back, is the previous version of the agent guaranteed to be next to the currently running agent in the data path? That is is the state of the filesystem always something like:

```
data/
  elastic-agent-current/
  elastic-agent-next/
```

> If this is true, can we look to see if there's another agent in the data directory next to us to detect if there was a roll back?

Two folders will exist the moment we go past the `UPG_EXTRACTING` state and, yes, two folders will also exist when we are in the `UPG_ROLLBACK` state.  So I'm not sure if checking if the number of folders > 1 is sufficient to determine if we're about to upgrade or if we've just rolled back.

But I think there might be another solution to detecting if we're about to upgrade: the code that creates the Upgrade Marker file runs in the same process as the code that's watching the Upgrade Marker file for changes.  As such, the former code can communicate to the latter in memory that we're about to upgrade.  In the case of a rollback, this communication will not happen.  

Let me explore this solution in a separate PR as it's not strictly related to this PR here.

Reply:
> But I think there might be another solution to detecting if we're about to upgrade: the code that creates the Upgrade Marker file runs in the same process as the code that's watching the Upgrade Marker file for changes. As such, the former code can communicate to the latter in memory that we're about to upgrade. In the case of a rollback, this communication will not happen.

> Let me explore this solution in a separate PR as it's not strictly related to this PR here.

https://github.com/elastic/elastic-agent/pull/3837

Reply:
> The following states occur right before or after the upgrade marker file comes into existence and, as such, do get persisted to it: UPG_REPLACING, UPG_WATCHING, UPG_ROLLBACK. So if the Agent were to restart during one of these states, the upgrade details from the upgrade marker would be restored to the Coordinator state (and from there sent to Fleet).

What happens to the upgrade action if an upgrade is interrupted after it is started but before it is completed? Does the agent start over from the beginning? Does it acknowledge it as if the upgrade had happened if though it didn't? Does it never get acknowledged?

What does the watcher do if it starts up, sees an upgrade marker, but the version of the agent that is currently running isn't the version that should be running?

It would be surprising for a user to see an upgrade stuck in `UPG_REPLACING` for example. If the upgrade is essentially aborted by a host reboot then `UPG_ROLLBACK` could be considered the correct state. The `UPG_WATCHING` always clears itself when the watcher stops but I'm not sure what happens in this state in this situation today.

I think the ideal thing to happen in this situation is that the upgrade restarts from the beginning when the agent host system comes back online.

Reply:
> What happens to the upgrade action if an upgrade is interrupted after it is started but before it is completed? Does the agent start over from the beginning? Does it acknowledge it as if the upgrade had happened if though it didn't? Does it never get acknowledged?

Looking at the code, once the Upgrade Marker has been created, if the Agent restarts, it will acknowledge the upgrade with Fleet _even if the upgrade may not have completed_.  In fact, it's entirely possible that Agent acknowledges the upgrade with Fleet while the Upgrade Watcher is still running and then the Upgrade Watcher decides to roll back the Agent.  As far as Fleet is concerned, the upgrade would've been reported as successful, and then within 10 minutes, the previous version of Agent would start showing again without any explanation as to why.

With Upgrade Details, in the above rollback scenario, the Upgrade Watcher will record the state as `UPG_ROLLBACK` in the Upgrade Marker file, which will get picked up by the main Agent process and sent to Fleet.

> What does the watcher do if it starts up, sees an upgrade marker, but the version of the agent that is currently running isn't the version that should be running?

Again, looking at the code...

First, if the Upgrade Watcher was running before an interruption stopped/killed it, and then the Upgrade Watcher was restarted, it will exit immediately because the watcher lock file, `watcher.lock`, will still exist.  In this case, there are two possibilities as to which state will be reported in Upgrade Details to Fleet:
- if the Upgrade Marker contains Upgrade Details, the state recorded in it will be reported to Fleet. This should be the `UPG_REPLACING` state as it's the last state that's persisted to the Upgrade Marker before the old Agent's upgrade code restarts the new Agent.
- if, for some reason, the Upgrade Marker does not contain Upgrade Details and the version of the running Agent is the same as the previous version recorded in the Upgrade Marker, `UPG_ROLLBACK` will be reported to Fleet.

If the Upgrade Watcher wasn't running yet when the upgrade process was interrupted, the Upgrade Watcher will start monitoring the Agent regardless of what version of Agent is currently running or what version is recorded in the Upgrade Marker file.  In this case, `UPG_WATCHING` will be reported to Fleet.  If the watch succeeds, the Upgrade Details will stop being reported to Fleet; note that the version of Agent being reported to Fleet will already be the new one in this case.  If the watch fails and Agent has to be rolled back, `UPG_ROLLBACK` will be reported to Fleet.

> I think the ideal thing to happen in this situation is that the upgrade restarts from the beginning when the agent host system comes back online.

Agreed but I think restarting the upgrade from the beginning after a crash is beyond the scope of this PR so I've created https://github.com/elastic/elastic-agent/issues/3860 to track this improvement.

Reply:
Agree fixing this is outside the scope of this PR given your explanation.

I also think we acknowledge the upgrade too early since it isn't synchronized with the watcher but that is also out of scope.